sql/src/main/java/io/crate/execution/engine/join/HashInnerJoinBatchIterator.java

```diff
@@ -156,19 +158,26 @@ public boolean moveNext() {
     }
 
     private void recreateBuffer() {
-        blockSize = calculateBlockSize();
+        blockSize = calculateBlockSize(circuitBreaker, estimatedRowSizeForLeft, numberOfRowsForLeft);
         this.buffer = new IntObjectHashMap<>(this.blockSize);
         numberOfRowsInBuffer = 0;
     }
 
-    private int calculateBlockSize() {
+    @VisibleForTesting
+    static int calculateBlockSize(CircuitBreaker circuitBreaker,
+                                  long estimatedRowSizeForLeft,
+                                  long numberOfRowsForLeft) {
         // In case statistics are not yet available
         if (estimatedRowSizeForLeft <= 0 || numberOfRowsForLeft <= 0 || circuitBreaker.getLimit() == -1) {
             return DEFAULT_BLOCK_SIZE;
         }
 
         int blockSize = (int) ((circuitBreaker.getLimit() - circuitBreaker.getUsed()) / estimatedRowSizeForLeft);
-        return (int) Math.min(numberOfRowsForLeft, blockSize);
+        blockSize = (int) Math.min(numberOfRowsForLeft, blockSize);
+
+        // In case no mem available from circuit breaker then still allocate for 10 rows
+        // as a CircuitBreakerException will be thrown anyway during processing.
```

Comment:
could also be 0 when left rows is 0 and 0 would lead to a BufferAllocationException on map instantiation. in any case 0 must be avoided ;)

Reply:
No, the check above checks <= 0

Reply:
ah right, sry didn't see